File: ExodusImport/Source/ExodusImport/Private/JsonObjects/JsonTerrainDetailPrototype.cpp
#include "JsonImportPrivatePCH.h"
#include "JsonTerrainDetailPrototype.h"
#include "macros.h"

//#define JSON_ENABLE_VALUE_LOGGING

using namespace JsonObjects;

JsonTerrainDetailPrototype::JsonTerrainDetailPrototype(JsonObjPtr data){
	load(data);
}

void JsonTerrainDetailPrototype::load(JsonObjPtr data){
	JSON_GET_VAR(data, textureId);
	JSON_GET_VAR(data, detailPrefabId);
	JSON_GET_VAR(data, detailPrefabObjectId);

	JSON_GET_VAR(data, detailMeshId);
	JSON_GET_VAR(data, detailMeshMaterials);

	JSON_GET_VAR(data, bendFactor);
	JSON_GET_PARAM(data, dryColor, getColor);//Gamma?
	JSON_GET_PARAM(data, healthyColor, getColor);

	JSON_GET_VAR(data, minWidth);
	JSON_GET_VAR(data, minHeight);
	JSON_GET_VAR(data, maxHeight);
	JSON_GET_VAR(data, maxWidth);
	JSON_GET_VAR(data, noiseSpread);
	JSON_GET_VAR(data, renderMode);
	JSON_GET_VAR(data, usePrototypeMesh);

	billboardFlag = (renderMode == TEXT("GrassBillboard"));
	vertexLitFlag = (renderMode == TEXT("VertexLit"));
	grassFlag = (renderMode == TEXT("Grass"));
}

File: ExodusExport/Scripts/Editor/Exporter/JsonTerrainDetailPrototype.cs
using UnityEditor;
using UnityEngine;
using System.Collections.Generic;

namespace SceneExport{
	[System.Serializable]
	public class JsonTerrainDetailPrototype: IFastJsonValue{
		public DetailPrototype data;
		
		public ResId textureId = ResId.invalid;
		public ResId detailPrefabId = ResId.invalid;
		public ResId detailPrefabObjectId = ResId.invalid;
		
		public MeshIdData meshIdData = new MeshIdData();
		public void writeRawJsonValue(FastJsonWriter writer){
			writer.beginRawObject();
			writer.writeKeyVal("bendFactor", data.bendFactor);
			writer.writeKeyVal("dryColor", data.dryColor);
			writer.writeKeyVal("minWidth", data.minWidth);
			writer.writeKeyVal("maxWidth", data.maxWidth);
			writer.writeKeyVal("minHeight", data.minHeight);
			writer.writeKeyVal("maxHeight", data.maxHeight);
			writer.writeKeyVal("healthyColor", data.healthyColor);
			writer.writeKeyVal("noiseSpread", data.noiseSpread);
			writer.writeKeyVal("renderMode", data.renderMode.ToString());
			writer.writeKeyVal("usePrototypeMesh", data.usePrototypeMesh);
			
			writer.writeKeyVal("textureId", textureId);
			writer.writeKeyVal("detailMeshId", meshIdData.meshId);
			writer.writeKeyVal("detailMeshMaterials", meshIdData.meshMaterials);
			writer.writeKeyVal("detailPrefabId", detailPrefabId);
			writer.writeKeyVal("detailPrefabObjectId", detailPrefabObjectId);
			
			writer.endObject();
		}		
		
		public JsonTerrainDetailPrototype(DetailPrototype data_, ResourceMapper resMap){		
			data = data_;
			
			textureId = resMap.getTextureId(data.prototypeTexture);
			detailPrefabId = resMap.getRootPrefabId(data.prototype, true);
			detailPrefabObjectId = resMap.getPrefabObjectId(data.prototype, true);
			
			meshIdData = new MeshIdData(data.prototype, resMap);
			//meshId = resMap.getPrefabObjectId
		}
	}
}